### PR TITLE
[script][clean-leather] Refactor

### PR DESCRIPTION
**Pull request by asechrest**
This is an attempt to refactor and modernize `clean-leather`. This refactoring was done in concert with @Valherun.

The existing script worked in some circumstances but failed in others. **I want to be clear up-front that this refactoring _does_ change the behavior of the script and the order of the arg inputs and whether they are required.**  The problems with the script were as follows:

- The original script makes the storage container optional (e.g. the container specified to store cleaned hides/skins). The problem with this is that rawhide and cleaned hide and curing hide all respond to the same `tap`. Without specifying a storage container, the script isn't robust enough to handle parsing raw skins and cleaned or curing skins in the same container. Further, quirks of DR crafting mean that sometimes a skin that is fully cured and _should_ respond to the noun `leather` do not do so until interacted with in some way. Thus, rather than trying to make the script properly handle intermixed skins in the same container, we now _require_ a storage container to keep cleaned and curing skins separate from raw skins.
- The original script did not allow the flexibility to apply preservative 1 time or max times. It now defaults to 1 time, but allows an option to apply max applications of preservative. Max applications do not seem to be a consistent number of applications (I got 6 and 7 applications to reach max in a test), thus I have not offered fine-grained number of applications. Either one or max.
- The original script did not handle running out of preservative mid-application of preservative. Now it does.
- The original script was not able to handle scraping and applying preservative to "skins" because of the very strange way that DR parses input for each of those actions and perhaps specifically with a skin (vs. a hide). The difficulty of the DR parser caused me to now require an arg that specifies the "animal" the rawhide comes from. This allows us to use that animal arg to handle the DR parser strangeness by either using or not using the animal arg to scrape or apply preservative accordingly.
- I've opted to remove the `include` lines in favor of explicit calls to the proper modules (e.g. `DRCI.stow_hands`).
- The original script included superfluous `custom_require.call` items. Now removed.
- I've changed the script to NOT use instance variables where possible.
- I've changed the script to utilize `bput` and DRC helper methods where possible in lieu of fputs.
- I've added some more robust messaging feedback when the script can't do certain things, and exit the script accordingly.
- Various well-defined actions of the script have been compartmentalized into separate methods.
- The code is perhaps a touch less "efficient" in favor of more readability.
- I've added comments to explain primary code blocks.

I welcome comments or testers. @Valherun has tested this with me and it appears to function, but again with the caveat that the arg order and whether required or not is changed from the original. I considered this the best course of action without having to make this script more dynamic and robust than perhaps is warranted for it.

### CODE REVIEW

**Comment by asechrest** [comment]
Example of DR crafting hiccup where certain items have the wrong description until interacted with in a certain way:

```
>rumm rucks
You rummage through a rucksack fastened with wooden leaves and see some bleaching jackal bones, some bleaching jackal bones and some bleaching jackal bones.
>get bones
What do you want to get?
>
[drinfomon]>exp all
>
[drinfomon]>info
  [You can pay off this debt in person at the respective provincial debt office or by calling for an urchin runner with BANK DEBT.]

>look in rucks
In the rucksack you see a jackal-bone stack, a jackal-bone stack and some bleaching jackal bones.
```

**Comment by asechrest** [comment]
Examples of the DR parser being screwy with skin scraping:

```
>get my skin
You get a kobold skin from inside your backpack.
>get my scraper
You get a hide scraper from inside your backpack.
>scrape my skin with my scraper
How do you want to clean this skin?
  SCRAPE (SKIN/PELT/HIDE) WITH SCRAPER (QUICK/NORMAL/CAREFUL)

>scrape my skin with my scraper quick
How do you want to clean this skin?
  SCRAPE (SKIN/PELT/HIDE) WITH SCRAPER (QUICK/NORMAL/CAREFUL)

>scrape my skin with scraper quick
How do you want to clean this skin?
  SCRAPE (SKIN/PELT/HIDE) WITH SCRAPER (QUICK/NORMAL/CAREFUL)

>scrape skin
You scrape your skin, cleaning some dirt and debris from it.
Roundtime: 14 sec.
>scrape my skin
You scrape your skin, cleaning some dirt and debris from it.
Roundtime: 11 sec.
>scrape my skin quick
What did you want to scrape?
>scrape skin with my scraper quick
You quickly scrape your skin, cleaning some dirt and debris from it.
The skin looks as clean as you can make it.
Roundtime: 1 sec.
```

**Comment by asechrest** [comment]
Example of the DR parser being even more ridiculous applying lotion to a skin. Note that the only combination that I feel could consistently work is `pour my lotion on my <animal> skin`. Also note that this is _different_ than the most consistent way to handle scraping the skin, which is _without_ the <animal> part. 

```
>stow l
You put your scraper in your backpack.
>get my lotion
You get some tanning lotion from inside your backpack.
>pour my lotion on my skin
What do you want to pour the lotion on?
>pour lotion on my skin
What do you want to pour the lotion on?
>pour lotion on skin
You pour some of the salty solution onto the skin and it begins to absorb into the material.
Roundtime: 15 sec.
>stow l
You put your lotion in your backpack.
>get other skin
What were you referring to?
>get skin from back
You get a kobold skin from inside your backpack.
>empty left
You drop a kobold skin.
>get lotion
You get some tanning lotion from inside your backpack.
>pour lotion on skin
You must be holding the kobold skin to do that.
>pour my lotion on skin
You must be holding the kobold skin to do that.
>pour my lotion on my skin
What do you want to pour the lotion on?
>pour my lotion on my kobold skin
You pour an addtional dose of salty solution onto the skin and vigorously rub it in to hasten the curing process.
Roundtime: 11 sec.
```

**Comment by asechrest** [comment]
Example of the original script struggling to keep track of rawhide and cleaned skins if both the source container and storage container are the same. Because they all respond to "skin" and don't respond to the proper adjective whether raw, cleaned, or curing, and they change order as we're stowing cleaned skins in the container and trying to get out uncleaned skins, we either need a super robust process to figure out how to get the right skins out of the container, or we just accept that we must have a storage container specified that is separate from our source container.

```
[clean-leather]>put my tanning lotion in my backpack
You put your lotion in your backpack.
>
[clean-leather]>stow my skin
You put your skin in your backpack.
>
[clean-leather]>get skin from my backpack
You get a cleaned prereni skin from inside your backpack.
>
[clean-leather]>get my scraper
You get a hide scraper from inside your backpack.
>
[clean-leather]>scrape skin with my scraper 
The skin looks as clean as you can make it!
>
[clean-leather]>put my scraper in my backpack
You put your scraper in your backpack.
```

File: clean-leather.lic [hunk 1]
```diff
@@ -2,80 +2,156 @@
...
+  def apply_preservative(animal, hide, storage, full_preservative)
+
+    # This loop applies a single application of preservative unless
+    # the user has specified "full" application in the args, in which case
+    # it applies until game messages suggest we're at max.
+    loop do
+      # If we have no preservative on-hand, get money,
+      # get preservative, go back to crafting spot.
+      unless DRCI.get_item_if_not_held?(@preservative)
+        DRCM.ensure_copper_on_hand(2000, @settings)
```
Comment (rpherbig) [inline clean-leather.lic:127]: `common-money` needs added to `custom_require`

File: clean-leather.lic [hunk 2]
```diff
@@ -2,80 +2,156 @@
...
+        break if !full_preservative
+        next
+      when /^Pour what?/i
+        # We've run out of lotion during the pouring process, re-run method, which will buy more
+        apply_preservative(animal, hide, storage, full_preservative)
+      when /would only be damaged by adding more/i
+        break
       else
-        fput("stow my #{args.noun}")
+        DRC.message("Unable to apply preservative. Please report this error.")
```
Comment (rpherbig) [inline clean-leather.lic:145]: Can we provide more guidance on _how_ to report this error?

File: clean-leather.lic (hunk 1 above)
Comment (asechrest) [reply clean-leather.lic:127]: Great catch. Thanks.

File: clean-leather.lic (hunk 2 above)
Comment (asechrest) [reply clean-leather.lic]: Done.

**Comment by asechrest** [comment]
@rpherbig 

Appreciate the review. Good comments.

**Pull request merged**